packages/snaps-utils/src/ui.tsx

```diff
@@ -0,0 +1,411 @@
+ * Get the text child from a list of markdown tokens.
+ *
+ * @param tokens - The markdown tokens.
+ * @returns The text child.
+ */
+function getTextChildFromTokens(tokens: Token[]) {
+  return getChildren(tokens.flatMap(getTextChildFromToken));
+}
+
+/**
+ * Get the text child from a markdown token.
+ *
+ * @param token - The markdown token.
+ * @returns The text child.
+ */
+function getTextChildFromToken(token: Token): TextChildren {
+  switch (token.type) {
+    case 'link': {
+      return <Link href={token.href} children={getLinkText(token)} />;
+    }
+
+    case 'text':
+      return token.text;
+
+    case 'strong':
+      return (
+        <Bold>
+          {
+            getTextChildFromTokens(
+              // Due to the way `marked` is typed, `token.tokens` can be
+              // `undefined`, but it's a required field of `Tokens.Bold`, so we
+              // can safely cast it to `Token[]`.
+              token.tokens as Token[],
+            ) as BoldChildren
+          }
+        </Bold>
+      );
+
+    case 'em':
+      return (
+        <Italic>
+          {
+            getTextChildFromTokens(
+              // Due to the way `marked` is typed, `token.tokens` can be
+              // `undefined`, but it's a required field of `Tokens.Bold`, so we
+              // can safely cast it to `Token[]`.
+              token.tokens as Token[],
+            ) as ItalicChildren
+          }
+        </Italic>
+      );
+
+    default:
+      return null;
+  }
+}
+
+/**
+ * Get all text children from a markdown string.
+ *
+ * @param value - The markdown string.
+ * @returns The text children.
+ */
+export function getTextChildren(value: string) {
+  const rootTokens = lexer(value, { gfm: false });
+  const children: TextChildren = [];
+
+  walkTokens(rootTokens, (token) => {
+    if (token.type === 'paragraph') {
+      if (children.length > 0) {
+        children.push('\n\n');
+      }
+
+      const { tokens } = token as Tokens.Paragraph;
+      children.push(...tokens.flatMap(getTextChildFromToken));
+    }
+  });
+
+  return children.filter((child) => child !== null);
+}
+
+/**
+ * Validate the text size of a component. The text size is the total length of
+ * all text in the component.
+ *
+ * @param component - The component to validate.
+ * @throws An error if the text size exceeds the maximum allowed size.
+ */
+function validateComponentTextSize(component: Component) {
+  const textSize = getTotalTextLength(component);
+  assert(
+    textSize <= MAX_TEXT_LENGTH,
+    `The text in a Snap UI may not be larger than ${
+      MAX_TEXT_LENGTH / 1000
+    } kB.`,
+  );
+}
+
+/**
+ * Get a JSX element from a legacy UI component. This supports all legacy UI
+ * components, and maps them to their JSX equivalents where possible.
+ *
+ * @param component - The legacy UI component.
+ * @returns The JSX element.
+ */
+export function getJsxElementFromComponent(component: Component): JSXElement {
+  validateComponentTextSize(component);
```

Comment:
`validateComponentTextSize` is already recursive, so we shouldn't call it at the top of this function as `getJsxElementFromComponent` is also used recursively.